### PR TITLE
server,mysql: support server-side cursors

### PR DESCRIPTION
**Pull request by mz1999**
fix https://github.com/pingcap/tidb/issues/6231
Implement server-side cursors by handling [COM_STMT_FETCH](https://dev.mysql.com/doc/internals/en/com-stmt-fetch.html) command.
The client indicates that it wants to use cursor
by setting a flag in [COM_STMT_EXECUTE](https://dev.mysql.com/doc/internals/en/com-stmt-execute.html) command
Subsequently, the client acquires result rows repeatedly by sending COM_STMT_FETCH,
which will carry stmt-id and fetch size.
This commit only Implements forward-only, read-only cursor

### CODE REVIEW

**Comment by shenli** [comment]
@mz1999 Thanks! Please sign the CLA.

File: mysql/const.go [hunk 1]
```diff
@@ -55,6 +55,11 @@ const (
 	ServerPSOutParams              uint16 = 0x1000
 )
 
+// IsCursorExists return true if cursor exists indicated by server status
+func IsCursorExists(serverStatus uint16) bool {
```
Comment (coocood) [inline mysql/const.go]: We have functions like this in this form:
```
func HasCursorExistsFlag(serverStatus uint16) bool {
    serverStatus&ServerStatusCursorExists > 0
}
```

And I think this function is too simple to write a test for it.

File: server/driver_tidb.go [hunk 2]
```diff
@@ -122,6 +123,7 @@ func (ts *TiDBStatement) Close() error {
 		return errors.Trace(err)
 	}
 	delete(ts.ctx.stmts, int(ts.id))
+	delete(ts.ctx.results, int(ts.id))
```
Comment (coocood) [inline server/driver_tidb.go]: The results should be Closed, or there may be goroutine leak.

Comment (coocood) [inline server/driver_tidb.go]: And we also need to Close all the results in connection close to avoid goroutine leak.

**Comment by winkyao** [comment]
ref COM_STMT_EXECUTE binary protocol https://mariadb.com/kb/en/library/com_stmt_execute/

File: mysql/const.go (hunk 1 above)
Comment (mz1999) [reply mysql/const.go]: It has been updated as required.

File: server/driver_tidb.go (hunk 2 above)
Comment (mz1999) [reply server/driver_tidb.go]: Updates have been committed.

Closing ResultSet to release resources when Statement or connection is closed.

File: mysql/errcode.go [hunk 3]
```diff
@@ -881,7 +881,8 @@ const (
 	ErrAlterOperationNotSupportedReasonNotNull                      = 1861
 	ErrMustChangePasswordLogin                                      = 1862
 	ErrRowInWrongPartition                                          = 1863
-	ErrErrorLast                                                    = 1863
+	ErrUnknownStmtFetchResultSet                                    = 1864
```
Comment (winkyao) [inline mysql/errcode.go]:
```
Error: 1864 SQLSTATE: HY000 (ER_MTS_EVENT_BIGGER_PENDING_JOBS_SIZE_MAX)

Message: Cannot schedule event %s, relay-log name %s, position %s to Worker thread because its size %lu exceeds %lu of slave_pending_jobs_size_max.
```
errcode 1864 is not the code return by MySQL, MySQL will return `1243` in this case

```
#define ER_UNKNOWN_STMT_HANDLER 1243

  if (!(stmt= thd->stmt_map.find(stmt_id)))
  {
    char llbuf[22];
    my_error(ER_UNKNOWN_STMT_HANDLER, MYF(0), static_cast<int>(sizeof(llbuf)),
             llstr(stmt_id, llbuf), "mysqld_stmt_fetch");
    DBUG_VOID_RETURN;
  }
```
You can find the error info in https://dev.mysql.com/doc/refman/5.7/en/error-messages-server.html

File: server/driver.go [hunk 4]
```diff
@@ -70,6 +70,12 @@ type QueryCtx interface {
 	// GetStatement gets PreparedStatement by statement ID.
 	GetStatement(stmtID int) PreparedStatement
 
+	// StoreResultSet stores ResultSet for subsequent using.
+	StoreResultSet(stmtID int, rs ResultSet)
```
Comment (coocood) [inline server/driver.go]: The `ResultSet` is associated with a `PreparedStatement`.
I think it is better to add the method to `PreparedStatement` instead of QueryCtx.

File: server/driver_tidb.go [hunk 5]
```diff
@@ -47,6 +48,7 @@ type TiDBContext struct {
 	session   session.Session
 	currentDB string
 	stmts     map[int]*TiDBStatement
+	results   map[int]ResultSet
```
Comment (coocood) [inline server/driver_tidb.go]: We can store the `ResultSet` in `TiDBStatment`.

File: server/conn.go [hunk 6]
```diff
@@ -990,13 +1005,17 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
 				return errors.Trace(err)
 			}
 		}
+		// if cursor exists, checking whether this batch has fetched enough rows.
+		if mysql.HasCursorExistsFlag(serverStatus) && fetchRows >= fetchSize {
```
Comment (coocood) [inline server/conn.go]: Is it ok to return rows more than `fetchSize`?

Comment (mz1999) [reply server/conn.go]: The client uses cursor for getting results in batches, instead of acquiring the whole result set at once. The size of each batch is controllable to avoid client Out of memory.

The client is still traversing the result set as if there is no cursor. It doesn't care about the fetch size, as long as it does not cause the client OOM.
So I think the current implementation is ok.

Of course, if the fetch size is strictly equal to the value specified by the client, it's a better implementation.
However, I think it is acceptable to meet the essential requirement with minimal modification.

Comment (coocood) [reply server/conn.go]: I worried about that some of the client has an assumption that the returned result set is no more than fetch size, then it will discard some rows in the result set.

Comment (coocood) [reply server/conn.go]: If we support strict fetch size, then the change would be too much, then we better define another function without modifying the `writeChunks` in case the change introduces bugs.

Comment (mz1999) [reply server/conn.go]: How about the following solution:

if `fetch size` < `MaxChunkSize`, then `init Chunk capacity` = `fetch size`
if `fetch size` >= `MaxChunkSize`, then `init Chunk capacity` = `MaxChunkSize`

Every time we handle the `COM_STMT_FETCH` command，we fill the chunk with rows only once. 
So the result size must be <= `fetch size`

**Comment by shenli** [comment]
@winkyao @coocood @lysu PTAL

File: server/conn.go (hunk 6 above)
Comment (coocood) [reply server/conn.go]: The `init chunk capacity` doesn't matter, the number of rows in a chunk is `MaxChunkSize`.

Comment (mz1999) [reply server/conn.go]: I notice `maxChunkSize` is the property of `baseExecutor`
 
Assuming I add a new method:

`NewChunk(chunkSize)`

It returns a new Chunk with init chunkSize, and setting the current baseExecutor's `maxChunkSize =  chunkSize` .  Doing this will not affect other executors.

how about this way?  Could you please give me some suggestions?

Comment (coocood) [reply server/conn.go]: I think this feature is not widely used, we can accept this feature only if it doesn't affect the existing logic.

We don't want this feature to add complexity to our normal execution code.

I think the best way is to define another method like `writeChunksWithFetchSize`,  the old `writeChunks` is not touched. 

Common code can be extracted to method though, that is ok.

Comment (lysu) [reply server/conn.go]: I think @coocood means "TiDB still read data as chunk, but flush output to client use `fetchSize`"? and TiDB hold a partial result buffer when chunk > fetchSize ????

Comment (lysu) [reply server/conn.go]: @mz1999 PTAL, modify chunk/executor is prefect solution but maybe need too big change.

PS: maybe add some unit test for this modify, current golang are not support server sider cursor integration test seem be not easy, thx~

Comment (coocood) [reply server/conn.go]: @mz1999 Great!

File: server/conn.go [hunk 7]
```diff
@@ -976,27 +988,72 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
...
 	}
-	return errors.Trace(cc.writeEOF(more))
+	return errors.Trace(cc.writeEOF(serverStatus))
+}
+
+// writeChunksWithFetchSize writes data from a Chunk, which filled data by a ResultSet, into a connection.
+// binary specifies the way to dump data. It throws any error while dumping data.
+// serverStatus, a flag bit represents server information
+// fetchSize, the desired number of rows to be fetched each time when client uses cursor
+func (cc *clientConn) writeChunksWithFetchSize(ctx context.Context, rs ResultSet, binary bool, serverStatus uint16, fetchSize int) error {
```
Comment (coocood) [inline server/conn.go]: The `binary` argument is always true.

File: server/conn.go [hunk 8]
```diff
@@ -976,27 +988,72 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
...
+		rowCount := chk.NumRows()
+		fetchRows += rowCount
+		if rowCount == 0 {
+			// tell the client COM_STMT_FETCH has finished by setting proper serverStatus,
+			// and close ResultSet
+			serverStatus |= mysql.ServerStatusLastRowSend
+			terror.Call(rs.Close)
+			break
+		}
+		err = cc.writeRowsInChunk(chk, data, rs.Columns(), binary)
```
Comment (coocood) [inline server/conn.go]: It may write more rows than `fetchSize`.

File: mysql/const.go [hunk 9]
```diff
@@ -55,6 +55,11 @@ const (
 	ServerPSOutParams              uint16 = 0x1000
 )
 
+// HasCursorExistsFlag return true if cursor exists indicated by server status
```
Comment (zimulala) [inline mysql/const.go]: s/retrun/retruns
Add `.` at the end of a line. Please update other similar places.

File: server/conn.go [hunk 10]
```diff
@@ -988,67 +988,77 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
...
-	fetchRows := 0
-	for {
+// serverStatus, a flag bit represents server information.
+// fetchSize, the desired number of rows to be fetched each time when client uses cursor.
+func (cc *clientConn) writeChunksWithFetchSize(ctx context.Context, rs ResultSet, serverStatus uint16, fetchSize int) error {
+	fetchedRows := rs.GetFetchedRows()
+	var curRows []chunk.Row
+
+	// if fetchedRows is empty, getting data from recordSet.
+	if len(fetchedRows) == 0 {
```
[the remark this answers is not in the document]
Comment (mz1999) [reply server/conn.go]: As long as the amount of rows returned to the client is '<=' fetch size, and the flag `ServerStatusCursorExists`  is set,  the client can process the data correctly, and knows that the data has not yet been sent completely, it will continue to send the `COM_STMT_FETCH` request for the next batch of rows.

For this bulk streaming operation, a batch of rows smaller than fetch size is acceptable as long as the overall result is correct.

[the remark this answers is not in the document]
Comment (mz1999) [reply server/conn.go]: You are right, let me try again

[the remark this answers is not in the document]
Comment (lysu) [reply server/conn.go]: Thanks. tommorrow I will do some test for this great job~ and try to push other guys to review this...thanks ^ ^

File: server/conn.go [hunk 11]
```diff
@@ -997,12 +1009,71 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
...
+// writeChunksWithFetchSize writes data from a Chunk, which filled data by a ResultSet, into a connection.
+// binary specifies the way to dump data. It throws any error while dumping data.
+// serverStatus, a flag bit represents server information.
+// fetchSize, the desired number of rows to be fetched each time when client uses cursor.
+func (cc *clientConn) writeChunksWithFetchSize(ctx context.Context, rs ResultSet, serverStatus uint16, fetchSize int) error {
+	fetchedRows := rs.GetFetchedRows()
+
+	// if fetchedRows is not enough, getting data from recordSet.
+	for len(fetchedRows) < fetchSize {
+		chk := rs.NewChunk()
```
[the remark this answers is not in the document]
Comment (mz1999) [reply server/conn.go]: I feel that this chunk cannot be reused. Because rows stored in fetchedRows refer to the underlying chunk.
If the chunk is reset, the data in Row is wrong.

[the remark this answers is not in the document]
Comment (lysu) [reply server/conn.go]: you are right~

**Comment by coocood** [comment]
@mz1999 
Thank you for your contribution!
